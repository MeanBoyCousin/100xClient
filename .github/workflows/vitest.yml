name: Vitest

on:
  pull_request:
    branches: [master]

  workflow_dispatch:

jobs:
  test:
    name: Test
    permissions:
      actions: write
      contents: read
    runs-on: ubuntu-latest
    timeout-minutes: 2

    steps:
      - name: Cancel old runs
        uses: styfle/cancel-workflow-action@0.12.0
        with:
          access_token: ${{ github.token }}

      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          cache: 'yarn'
          node-version: '18.x'

      - name: Install packages
        run: yarn

      - name: Run tests
        run: yarn test --silent
