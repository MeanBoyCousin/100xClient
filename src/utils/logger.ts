const logger = {
  failure: (msg: string) => console.error('\x1b[31m✖\x1b[0m', msg),
  message: (msg: string) => console.log('\x1b[33m→\x1b[0m', msg),
  success: (msg: string) => console.log('\x1b[32m✔\x1b[0m', msg),
}

export default logger
