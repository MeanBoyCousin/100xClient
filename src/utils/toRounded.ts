/**
 * Round a decimal number to x places.
 *
 * @param value - The number to be rounded.
 * @param [precision] (Optional) The amount of decimals to round to (default: 2).
 * @returns The rounded value.
 */
const toRounded = (value: number, precision: number = 2): number =>
  parseFloat(
    (Math.floor(value * Math.pow(10, precision)) / Math.pow(10, precision)).toPrecision(15),
  )

export default toRounded
