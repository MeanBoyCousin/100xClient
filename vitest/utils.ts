// User variables.
export const address = '0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8'
export const privateKey = '0xa608cd43cbc3d59bc51443f475b96a4654e956d6cc91783598a8e76a34000174'

// Contract addresses
export const ciaoAddress = '0x63bD0ca355Cfc117F5176E5eF3e34A6D60081937'
export const USDB = '0x79a59c326c715ac2d31c169c85d1232319e341ce'

// API response mocks
export const productsData = [
  {
    id: 1003,
    type: 'PERP',
    symbol: 'btcperp',
    active: true,
    baseAsset: 'WBTC',
    quoteAsset: 'USDB',
    minQuantity: '1000000000000000',
    maxQuantity: '1000000000000000000000000',
    increment: '100000000000000',
    takerFee: '500000000000000',
    makerFee: '300000000000000',
    isMakerRebate: true,
    initialLongWeight: '900000000000000000',
    initialShortWeight: '1100000000000000000',
    maintenanceLongWeight: '950000000000000000',
    maintenanceShortWeight: '1050000000000000000',
    baseAssetAddress: '0xdc2f16a474a969056e6a559629b46d01f8675a1a',
    quoteAssetAddress: '0x79a59c326c715ac2d31c169c85d1232319e341ce',
    markPrice: '72581488537710140836821',
  },
  {
    id: 1002,
    type: 'PERP',
    symbol: 'ethperp',
    active: true,
    baseAsset: 'WETH',
    quoteAsset: 'USDB',
    minQuantity: '100000000000000',
    maxQuantity: '100000000000000000000000',
    increment: '10000000000000',
    takerFee: '200000000000000',
    makerFee: '50000000000000',
    isMakerRebate: true,
    initialLongWeight: '950000000000000000',
    initialShortWeight: '1050000000000000000',
    maintenanceLongWeight: '970000000000000000',
    maintenanceShortWeight: '1030000000000000000',
    baseAssetAddress: '0x4200000000000000000000000000000000000023',
    quoteAssetAddress: '0x79a59c326c715ac2d31c169c85d1232319e341ce',
    markPrice: '3399999999999999999772',
  },
]

export const klines = [
  {
    e: 'klines_',
    E: 1710349200031,
    s: 'ethperp',
    i: '1h',
    o: '4000000000000000000000',
    c: '4000000000000000000000',
    h: '4000000000000000000000',
    l: '4000000000000000000000',
    v: '0',
    x: true,
  },
  {
    e: 'klines_',
    E: 1710352800031,
    s: 'ethperp',
    i: '1h',
    o: '0',
    c: '3985200000000000000000',
    h: '3989000000000000000000',
    l: '3983700000000000000000',
    v: '5940000000000000000',
    x: true,
  },
  {
    e: 'klines_',
    E: 1710356400031,
    s: 'ethperp',
    i: '1h',
    o: '3985200000000000000000',
    c: '3977300000000000000000',
    h: '4005000000000000000000',
    l: '3962700000000000000000',
    v: '66960000000000000000',
    x: true,
  },
  {
    e: 'klines_',
    E: 1710360000031,
    s: 'ethperp',
    i: '1h',
    o: '3977300000000000000000',
    c: '3994800000000000000000',
    h: '4005400000000000000000',
    l: '3976000000000000000000',
    v: '4850000000000000000',
    x: true,
  },
  {
    e: 'klines_',
    E: 1710363600031,
    s: 'ethperp',
    i: '1h',
    o: '3994800000000000000000',
    c: '3993800000000000000000',
    h: '4007700000000000000000',
    l: '3988600000000000000000',
    v: '3320000000000000000',
    x: true,
  },
]

export const tickers = [
  {
    productId: 1002,
    productSymbol: 'ethperp',
    oraclePrice: '3633138612183610451397',
    markPrice: '3378124999999999999818',
    priceChange: '10000000000000000000000000',
    priceChangePercent: '0',
    volume: '3621652000000000000000',
    low: '0',
    high: '10000000000000000000000000',
    fundingRateHourly: '-499999999999999900',
    fundingRateYearly: '-4382999999999999123400',
    openInterest: '433926912499999999976621',
    nextFundingTime: 1712667600,
  },
  {
    productId: 1003,
    productSymbol: 'btcperp',
    oraclePrice: '70816037672632469474700',
    markPrice: '70806062729901863024587',
    priceChange: '75113700000000000000000',
    priceChangePercent: '0',
    volume: '96050000000000000000',
    low: '0',
    high: '75113700000000000000000',
    fundingRateHourly: '-10848157239100',
    fundingRateYearly: '-95094946357950600',
    openInterest: '229411643244882036199661',
    nextFundingTime: 1712667600,
  },
]
