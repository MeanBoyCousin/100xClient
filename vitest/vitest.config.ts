import { resolve } from 'path'

import { defineConfig } from 'vitest/config'

const config = defineConfig({
  resolve: {
    alias: [
      { find: 'src', replacement: resolve(__dirname, '../', './src') },
      { find: 'vitest/utils', replacement: resolve(__dirname, '../', './vitest/utils.ts') },
    ],
  },
  root: './src',
  test: {
    coverage: {
      all: true,
      exclude: [
        '__mocks__/',
        '__tests__/',
        'types/',
        '**/types.ts',
        '**/*.d.ts',
        'constants/**',
        'ABI/**',
        'sandbox.ts',
      ],
      reporter: ['html', 'lcov'],
      reportsDirectory: '../reports/vitest',
      watermarks: {
        statements: [80, 95],
        functions: [80, 95],
        branches: [80, 95],
        lines: [80, 95],
      },
    },
    environment: 'node',
    globals: true,
    setupFiles: '../vitest/vitest.setup.ts',
  },
})

export default config
