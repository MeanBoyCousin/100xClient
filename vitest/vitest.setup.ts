import { vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

const fetchMocker = createFetchMock(vi)
fetchMocker.enableMocks()

Object.defineProperty(BigInt.prototype, 'toJSON', {
  get() {
    'use strict'
    return () => String(this)
  },
})
