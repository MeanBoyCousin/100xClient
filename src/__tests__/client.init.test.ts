import { beforeEach, describe, expect, it, vi } from 'vitest'

import HundredXClient from 'src'
import { privateKey } from 'vitest/utils'

describe('The HundredXClient', () => {
  beforeEach(() => {
    fetchMock.resetMocks()
  })

  it('should initialise correctly with only required params', () => {
    const Client = new HundredXClient(privateKey)

    expect(Client).toMatchSnapshot()
  })

  it('should initialise correctly with all config parameters passed', () => {
    const Client = new HundredXClient(privateKey, {
      debug: true,
      environment: 'testnet',
      rpc: 'https://test-rpc.quiknode.pro',
      subAccountId: 2,
    })

    expect(Client).toMatchSnapshot()
  })

  it('should POST the referral code if running on mainnet', async () => {
    fetchMock.mockResponse(JSON.stringify({ success: true }))

    new HundredXClient(privateKey, { environment: 'mainnet' })

    await vi.waitFor(() => {
      if (!fetchMock.mock.calls.length) throw new Error('Awaiting call...')
    })

    expect(fetchMock.mock.calls[0]).toMatchInlineSnapshot(`
      [
        "https://api.100x.finance/v1/referral/add-referee",
        {
          "body": "{"account":"0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8","code":"kickflip","signature":"0x5da080cbf906558a2642b7642572b3a5c75966e2bf3f9a8c5b86e77ef7b6395c3557266aa6971790d65f3c791d68e7894285f18bf16cecf5b9b06b53715fc1d21b"}",
          "method": "POST",
        },
      ]
    `)
  })

  it('should handle referral failures gracefully', async () => {
    fetchMock.mockReject(new Error('An unknown error occurred'))

    const Client = new HundredXClient(privateKey, { debug: true, environment: 'mainnet' })

    await vi.waitFor(() => {
      if (Client.logs.length !== 4) throw new Error('Awaiting call...')
    })

    expect(Client.logs[3]).toEqual({ msg: 'Call failed, ignoring.' })
  })
})
