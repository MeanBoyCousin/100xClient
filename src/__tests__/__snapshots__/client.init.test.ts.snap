// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The HundredXClient > should initialise correctly with all config parameters passed for a mainnet setup 1`] = `
HundredXClient {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "chain": 81457,
  "ciaoAddress": "0x0",
  "deposit": [Function],
  "domain": {
    "chainId": 81457n,
    "name": "100x",
    "verifyingContract": "0x0",
    "version": "0.0.0",
  },
  "environment": "mainnet",
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "logs": [
    {
      "msg": "Debug mode enabled",
    },
  ],
  "privateKey": "0xa608cd43cbc3d59bc51443f475b96a4654e956d6cc91783598a8e76a34000174",
  "rpc": "https://test-rpc.quiknode.pro",
  "subAccountId": 2,
  "verifierAddress": "0x0",
  "withdraw": [Function],
}
`;

exports[`The HundredXClient > should initialise correctly with all config parameters passed for a testnet setup 1`] = `
HundredXClient {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "chain": 168587773,
  "ciaoAddress": "0x63bD0ca355Cfc117F5176E5eF3e34A6D60081937",
  "deposit": [Function],
  "domain": {
    "chainId": 168587773n,
    "name": "100x",
    "verifyingContract": "0x65CbB566D1A6E60107c0c7888761de1AdFa1ccC0",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "logs": [
    {
      "msg": "Debug mode enabled",
    },
  ],
  "privateKey": "0xa608cd43cbc3d59bc51443f475b96a4654e956d6cc91783598a8e76a34000174",
  "rpc": "https://test-rpc.quiknode.pro",
  "subAccountId": 2,
  "verifierAddress": "0x65CbB566D1A6E60107c0c7888761de1AdFa1ccC0",
  "withdraw": [Function],
}
`;

exports[`The HundredXClient > should initialise correctly with only required params 1`] = `
HundredXClient {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "chain": 168587773,
  "ciaoAddress": "0x63bD0ca355Cfc117F5176E5eF3e34A6D60081937",
  "deposit": [Function],
  "domain": {
    "chainId": 168587773n,
    "name": "100x",
    "verifyingContract": "0x65CbB566D1A6E60107c0c7888761de1AdFa1ccC0",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "logs": [],
  "privateKey": "0xa608cd43cbc3d59bc51443f475b96a4654e956d6cc91783598a8e76a34000174",
  "rpc": "https://sepolia.blast.io",
  "subAccountId": 1,
  "verifierAddress": "0x65CbB566D1A6E60107c0c7888761de1AdFa1ccC0",
  "withdraw": [Function],
}
`;
